Search Results: Add new filters panel

vovakulikov opened the pull request:
## What have been done in this PR
This PR contains changes about filter panel UI. Most of these changes were implemented by design spec [here](https://www.figma.com/file/EPgl0qVbqGSumvf52brhda/Improving-search-results?type=design&node-id=431-55452&mode=design&t=JVOoFT0yVIjLTXz3-0). In a nutshell 
- New filter panel UI 
- Context-free filters (like first section with type filters)
- Context-based filters (like repositories, commits authors and symbol kinds)
- Improvements about caching system so changing filters won't trigger query twice

https://github.com/sourcegraph/sourcegraph/assets/18492575/42f4c101-702c-4340-ad21-ab24c3c878e3


## Follow-ups
- Static filters, at the moment filters like "Symbols kind", "Commit date" and "Authors" are generated on the client and therefore they don't have count info and it also possible that some of them may produce zero state. This is because currently we don't have them in our Search API. 
- Polishing designs details such icons and badges (in this PR I was focusing more on how filters work rather how they look, polishing will be done in a follow up PR to simplify tech and design review)
- Mobile layout for the search results page with new filters UI
- Removing feature flags (some flags about search results page will go away very soon)

## Feature flags setup

This feature eventually will be enabled by default but at the moment it's behind a feature flag `newSearchResultFiltersPanel`, also make sure that your other feature flag `newSearchResultsUI` is enabled too since new filters UI is added only to the most recent version of the search results page (new search results page feature flag will be removed in a follow up PR) 

## Test plan
- Check that you can see and interact with new filters panel if you have enabled `newSearchResultFiltersPanel: true` feature flag
- Check that the old panel got no regressions

<!-- All pull requests REQUIRE a test plan: https://docs.sourcegraph.com/dev/background-information/testing_principles 

Why does it matter? 

These test plans are there to demonstrate that are following industry standards which are important or critical for our customers. 
They might be read by customers or an auditor. There are meant be simple and easy to read. Simply explain what you did to ensure 
your changes are correct!

Here are a non exhaustive list of test plan examples to help you:

- Making changes on a given feature or component: 
  - "Covered by existing tests" or "CI" for the shortest possible plan if there is zero ambiguity
  - "Added new tests" 
  - "Manually tested" (if non trivial, share some output, logs, or screenshot)
- Updating docs: 
  - "previewed locally" 
  - share a screenshot if you want to be thorough
- Updating deps, that would typically fail immediately in CI if incorrect
  - "CI" 
  - "locally tested" 
-->

jasonhawkharris commented on client/branded/src/search-ui/results/filters/components/filters-doc-footer/FiltersDocFooter.tsx:
```diff
@@ -0,0 +1,25 @@
+    className?: string
+}
+
+export const FiltersDocFooter: FC<FiltersDocFooterProps> = ({ className }) => (
```
It would be nice to change the positioning of this so that it is fixed to the bottom of the filter panel, regardless of scroll position. If the search produces so many results that it hides this, it might be frustrating for a user desiring to be more granular in their code search usage.

vovakulikov replied:
@jasonhawkharris my logic here is if we have a lot of filters content we probably should want to render as much them as possible to give more context and filtering options to the user. 

The filter footer section in my opinion is kind of helper here when you picked some filters but this produced a zero result state (when this footer panel will be visible and placed at the bottom of the filters panel) 

cc @taiyab

jasonhawkharris replied:
@taiyab I'm curious about your thoughts here.

taiyab replied:
@jasonhawkharris Basically what @vovakulikov said. The footer section is almost a "tried everything else and not got what you want, OK go here" type prompt. It doesn't need to be so prominently in someone's face, and it's likely someone will scroll to the bottom in their attempt to find whatever other options they're looking for.

jasonhawkharris commented on client/branded/src/search-ui/results/filters/NewSearchFilters.tsx:
```diff
@@ -0,0 +1,162 @@
+interface NewSearchFiltersProps {
+    query: string
+    results: SearchMatch[] | undefined
+    filters?: ResultFilter[]
```
Is there a time when this could be undefined/null?

vovakulikov replied:
Yes, when we just started observe results from the backend (right after user submitted their search) default value for filters is undefined in the stream search API.

jasonhawkharris commented on client/branded/src/search-ui/results/filters/components/dynamic-filter/SearchDynamicFilter.tsx:
```diff
@@ -0,0 +1,279 @@
+            )}
+
+            <ul className={styles.list}>
+                {filtersToShow.map(filter => {
```
readability nit: could this be extracted to its own function?

jasonhawkharris commented on client/branded/src/search-ui/results/filters/components/filter-type-list/FilterTypeList.tsx:
```diff
@@ -0,0 +1,157 @@
+            return 'diff'
+        }
+
+        default: {
```
Do we need the default case here? Would we ever pass anything other than repo, path, symbol, commit or diff into `toSearchSyntaxTypeFilter`?

vovakulikov replied:
Yeah because of Code type which is in my opinion isn't `type:file` but `type:file OR type:commit OR type:diffs`, since this filter type doesn't have one single types we don't assign any type filter to this section but rather just clear search query from all type filters.

jasonhawkharris commented on client/browser/src/shared/code-hosts/gitlab/scrape.ts:
```diff
@@ -54,6 +54,7 @@ export const getPageKindFromPathName = (owner: string, projectName: string, path
  */
 export const getGitlabRepoURL = (): string => {
     const projectLink = document.querySelector<HTMLAnchorElement>('.context-header a, .shortcuts-project')
+    console.log('link', projectLink)
```
I'm assuming this was for debugging and can be deleted?

fkling replied:
Delete

```suggestion
```

jasonhawkharris commented on client/web/src/repo/tree/TreePagePanels.tsx:
```diff
@@ -25,7 +27,7 @@ import type { BlobFileFields, TreeHistoryFields } from '../../graphql-operations
 import { RenderedFile } from '../blob/RenderedFile'
 import { CommitMessageWithLinks } from '../commit/CommitMessageWithLinks'
-import { FILE_ICONS, FileExtension, getFileInfo } from '../fileIcons'
+import { getFileInfo } from '../fileIcons'
```
The icon logic may need to be reworked when [this PR](https://github.com/sourcegraph/sourcegraph/pull/59234) is merged. You may want to wait to merge this until that one is merged, as it will probably change the invocations. For example, `getFileInfo` as a function won't even exist any more once it is merged.

jasonhawkharris replied:
Specifically, some language detection improvements changed quite a few aspects of how the file icons are populated. 

cc: @varungandhi-src

vovakulikov replied:
I left a comment about this in the Varun's PR, will merge his and then I adjust this one. Thanks for the heads up!

jasonhawkharris commented on client/wildcard/src/components/Icon/language-icon/LanguageIcon.module.scss:
```diff
@@ -0,0 +1,52 @@
+.icon {
```
See comment above regarding the file-icons PR. This might create some nasty merge conflicts with [this](https://github.com/sourcegraph/sourcegraph/pull/59234).

jasonhawkharris commented on schema/settings.schema.json:
```diff
@@ -13,6 +13,14 @@
       "type": "object",
       "additionalProperties": true,
       "properties": {
+        "newSearchResultFilters": {
```
May be good to note that the filters themselves haven't changed in terms of what's available, just the way in which we filter. 

```suggestion
        "newSearchResultFiltersPanel": {
          "description": "Enables new experimental search results filters panel"
```

jasonhawkharris reviewed (commented):
LGTM overall. A few nits, a few suggestions. 

I'm going to stop short of approving because in terms of language Icons, a lot of the logic may need to change to accommodate [this PR](https://github.com/sourcegraph/sourcegraph/pull/59234) when it is merged, which will be soon. 

I think this is a vast improvement on the product as a whole. Killer job getting it out so fast!

fkling commented on client/branded/src/search-ui/results/filters/types.ts:
```diff
@@ -0,0 +1,74 @@
+    ByCommitDate,
+    ByDiffType,
+}
+
+export const TYPES_TO_FILTERS = {
+    [SearchFilterType.Code]: [
+        SearchResultFilters.ByLanguage,
+        SearchResultFilters.ByRepository,
+        SearchResultFilters.ByPath,
+        SearchResultFilters.Recipes,
+        SearchResultFilters.ArchivedAndForked,
+    ],
+    [SearchFilterType.Repositories]: [
+        SearchResultFilters.ByLanguage,
+        SearchResultFilters.ByMetadata,
+        SearchResultFilters.ArchivedAndForked,
+    ],
+    [SearchFilterType.Paths]: [
+        SearchResultFilters.ByLanguage,
+        SearchResultFilters.ByRepository,
+        SearchResultFilters.ArchivedAndForked,
+    ],
+    [SearchFilterType.Symbols]: [
+        SearchResultFilters.BySymbolKind,
+        SearchResultFilters.ByRepository,
+        SearchResultFilters.ByPath,
+    ],
+    [SearchFilterType.Commits]: [
+        SearchResultFilters.ByAuthor,
+        SearchResultFilters.ByRepository,
+        SearchResultFilters.ByCommitDate,
+        SearchResultFilters.ArchivedAndForked,
+    ],
+    [SearchFilterType.Diffs]: [
+        SearchResultFilters.ByDiffType,
+        SearchResultFilters.ByAuthor,
+        SearchResultFilters.ByRepository,
+        SearchResultFilters.ArchivedAndForked,
+    ],
+}
```
I can't find where these are used... are they still needed?

fkling commented on client/shared/src/search/query/transformer.ts:
```diff
@@ -21,7 +21,7 @@ export function omitFilter(query: string, filter: Filter): string {
 }
 
-const succeedScan = (query: string): Token[] => {
+export const succeedScan = (query: string): Token[] => {
```
If we want to expose this function, it would make more sense to me to move it to `scanner.ts`.

fkling commented on client/shared/src/search/stream.ts:
```diff
@@ -270,7 +270,7 @@ export interface Filter {
     count: number
     limitHit: boolean
-    kind: 'file' | 'repo' | 'lang' | 'utility'
+    kind: 'file' | 'repo' | 'lang' | 'utility' | 'select' | 'after' | 'before' | 'author'
```
I find it questionable to extend the server side filter kind with client side values (that's what's happening here, isn't it?). If there is no other way I'd at least move these into a separate line and a comment about it.

vovakulikov replied:
It makes sense yeah, I created a separate interface for new filter panel components which extends the backend one. When backend supports new type of filters we could remove it from one place.

fkling commented on client/web/src/search/results/SearchResultsCacheProvider.tsx:
```diff
@@ -68,14 +91,15 @@ export function useCachedSearchResults(
             return merge(stream.pipe(throttleTime(500)), stream.pipe(last())).pipe(
                 tap(results => {
-                    cachedResults.current = { results, query, options }
+                    const previosCache = cachedResults.current?.cache ?? {}
+                    cachedResults.current = { query, options, cache: { ...previosCache, [filterQuery]: results } }
```
```suggestion
                    const previousCache = cachedResults.current?.cache ?? {}
                    cachedResults.current = { query, options, cache: { ...previousCache, [filterQuery]: results } }
```

fkling reviewed (commented):
I'll do another pass tomorrow.
Note that the file icon stuff is being changed in https://github.com/sourcegraph/sourcegraph/pull/59234 .

vovakulikov commented on client/branded/src/search-ui/results/filters/NewSearchFilters.module.scss:
```diff
@@ -0,0 +1,45 @@
+    gap: 0.5rem;
+    background-color: var(--code-bg);
+
+    &::-webkit-scrollbar {
+        width: 0.5rem;
+        height: 0.5rem;
+    }
+
+    &::-webkit-scrollbar-thumb {
+        background-color: var(--oc-gray-4);
+    }
+
+    &::-webkit-scrollbar-track {
+        background-color: var(--oc-gray-1);
+    }
+
+    @-moz-document url-prefix('') {
+        scrollbar-width: thin;
+        scrollbar-color: var(--oc-gray-1);
+    }
+
+    :global(.theme-dark) & {
+        &::-webkit-scrollbar-thumb {
+            background-color: var(--oc-gray-8);
+        }
+
+        &::-webkit-scrollbar-track {
+            background-color: var(--oc-gray-7);
+        }
+
+        @-moz-document url-prefix('') {
+            scrollbar-width: thin;
+            scrollbar-color: var(--oc-gray-6);
+        }
+    }
```
@fkling this is just hacks due to needs to set different group of rules in different browsers, to my understanding flat list of rules might cause problems because maybe at the moment these rules don't overlap but someday they will. 

I was thinking about improvements here anyway since I think we don't have a good solution at the moment in our codebase for unified scrollbars, everyone does this differently, I took this snippet from some place in Code Insights because I used it before. Ideally, we should have either separate component or CSS mixin for this kind of things. I will try to do it shortly after this PR.

vovakulikov commented:
@jasonhawkharris @fkling I resolved most of the comments and adjusted by the new lang icons approach, take a look whenever you have a minute, thanks

fkling reviewed (approved):
Did you merge in the other PR? Or why are there so many file icon related changes now? Regardless, you should rebase on top of https://github.com/sourcegraph/sourcegraph/pull/59268 to "fix" the Timestamp test.

Accepting to unblock.

jasonhawkharris commented on client/branded/src/search-ui/results/filters/NewSearchFilters.tsx:
```diff
@@ -0,0 +1,162 @@
+                />
+            )}
+
+            {(type === SearchFilterType.Commits || type === SearchFilterType.Diffs) && (
```
So, if I'm reading this correctly, if this expression evaluates to true, this means that 'By author' and 'By commit date' will render above the other dynamic filters like 'By language', 'by repositories' and 'by file'. 

IMO, 'By author' and 'By commit date', being less common use cases, should be positioned below these other more common filter types.

vovakulikov replied:
By author and by commit date are more specific when you select type:commit or type:diff filters IMO. But to be honest I don't have any preference here just got it order from the Figma design.

taiyab replied:
@jasonhawkharris You defo have a point, but I also didn't want the list to look the same across different types.

It should be immediately obvious you can filter different things across different types, along with the fact that `By author` and `By commit date` are more specific to that type.

jasonhawkharris reviewed (approved):
Nice! Killer job. Very excited about this.

vovakulikov commented:
> Did you merge in the other PR? Or why are there so many file icon related changes now? 

I rebased the main and fixed the language icon component, the problem with new approach was that it newly added langs icon logic wasn't straightforward if you want to reuse the language icons. So my idea was to create a common LanguageIcon component where you can pass just lang name (extension optionally) and render lang icon. But to do this I had to refactor lang icon logic and also resoled common → wildcard → web dependencies. So my point is I had to do some changes about lang icons to reuse them in filters UI. 

I rebased onto the main with timestamp fix so I think this should be ok now.

taiyab commented:
Great job on this @vovakulikov!
